### ISSUE TITLE
Explicit non-'static lifetimes ignored in explicit closure argument types

### ISSUE THREAD

**Issue by mcy** [rust-lang/rust#87290]
I tried this code: https://godbolt.org/z/Pdj46T1v5

```rust
pub fn f<'a>(_: &'a i32, _: impl FnOnce(&mut &mut &'a i32)) {}

pub fn g<'a>(p: &'a i32) {
    f(p, |_: &mut &mut &'a i32| {})
}
```

I expected an error along the lines of "expected `&mut &mut &i32`, got `&mut &mut &'a i32`." Instead, rustc accepts this program.
The following modification does *not* compile with the expected error.
```rust
pub fn f<'a>(_: &'a i32, _: impl FnOnce(&mut &mut &'a i32)) {}

pub fn g<'a>(p: &'a i32) {
    f(p, |x: &mut &mut &'a i32| {
        let _: &mut &mut &'a i32 = x;
    })
}
```

**Comment by Enselic** [comment]
Triage: This compiles in 1.63 but not 1.62. It would be interesting to run https://github.com/rust-lang/cargo-bisect-rustc on the corresponding nightly toolchains to find what commit that fixed it, but all that's strictly needed is to add a regression test. I looked for an existing one but couldn't find one.
